driver-core/src/main/com/mongodb/internal/connection/BaseCluster.java

```diff
@@ -302,55 +309,64 @@ private boolean handleServerSelectionRequest(final ServerSelectionRequest reques
-     * unpicked elements left.
-     * <p>
-     * Note that this method may reorder the {@code list}, as it uses the
-     * <a href="https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle">Fisher–Yates, a.k.a. Durstenfeld, shuffle algorithm</a>.
-     */
-    private static List<ServerTuple> atMostNRandom(final ArrayList<ServerDescription> list, final int n,
-            final Function<ServerDescription, ServerTuple> transformer) {
-        ThreadLocalRandom random = ThreadLocalRandom.current();
-        List<ServerTuple> result = new ArrayList<>(n);
-        for (int i = list.size() - 1; i >= 0 && result.size() < n; i--) {
-            Collections.swap(list, i, random.nextInt(i + 1));
-            ServerTuple serverTuple = transformer.apply(list.get(i));
-            if (serverTuple != null) {
-                result.add(serverTuple);
-            }
-        }
-        return result;
-    }
-
-    private ServerSelector getCompositeServerSelector(final ServerSelector serverSelector) {
-        ServerSelector latencyMinimizingServerSelector =
-                new LatencyMinimizingServerSelector(settings.getLocalThreshold(MILLISECONDS), MILLISECONDS);
-        if (settings.getServerSelector() == null) {
-            return new CompositeServerSelector(asList(serverSelector, latencyMinimizingServerSelector));
-        } else {
-            return new CompositeServerSelector(asList(serverSelector, settings.getServerSelector(), latencyMinimizingServerSelector));
-        }
+    private static ServerSelector getCompleteServerSelector(
+            final ServerSelector serverSelector,
+            final ServerDeprioritization serverDeprioritization,
+            final ServersSnapshot serversSnapshot,
+            final ClusterSettings settings) {
+        // The set of `Server`s maintained by the `Cluster` is updated concurrently with `clusterDescription` being read.
+        // Additionally, that set of servers continues to be concurrently updated while `serverSelector` selects.
+        // This race condition means that we are not guaranteed to observe all the servers from `clusterDescription`
+        // among the `Server`s maintained by the `Cluster`.
+        // To deal with this race condition, we take `serversSnapshot` of that set of `Server`s
+        // (the snapshot itself does not have to be atomic) non-atomically with reading `clusterDescription`
+        // (this means, `serversSnapshot` and `clusterDescription` are not guaranteed to be consistent with each other),
+        // and do pre-filtering to make sure that the only `ServerDescription`s we may select,
+        // are of those `Server`s that are known to both `clusterDescription` and `serversSnapshot`.
+        // This way we are guaranteed to successfully get `Server`s from `serversSnapshot` based on the selected `ServerDescription`s.
+        //
+        // The pre-filtering we do to deal with the race condition described above is achieved by this `ServerSelector`.
```

Comment:
I think this comment should be in the code, not somewhere else, because it explains the code, that is not that obvious otherwise.

> but now that there is a solution

Note that it has been solved previously, i.e., it is solved in the `master`, just differently.

> all the facts mentioned in this comment seem like they should be evident to unfamiliar readers

This is a huge overstatement in my opinion.

Reply:
> there should be a test that ensures items missing from the snapshot get filtered out by this chain

Done in https://github.com/mongodb/mongo-java-driver/pull/1367/commits/246353f36dae51e28aa212488444ec54cb8ad5af.